fuse_models/src/omnidirectional_3d.cpp

```diff
@@ -0,0 +1,629 @@
+    state1.acc_linear.z();
+
+  position2->data()[fuse_variables::Position3DStamped::X] = state2.position.x();
+  position2->data()[fuse_variables::Position3DStamped::Y] = state2.position.y();
+  position2->data()[fuse_variables::Position3DStamped::Z] = state2.position.z();
+
+  orientation2->data()[fuse_variables::Orientation3DStamped::X] = state2.orientation.x();
+  orientation2->data()[fuse_variables::Orientation3DStamped::Y] = state2.orientation.y();
+  orientation2->data()[fuse_variables::Orientation3DStamped::Z] = state2.orientation.z();
+  orientation2->data()[fuse_variables::Orientation3DStamped::W] = state2.orientation.w();
+
+  velocity_linear2->data()[fuse_variables::VelocityLinear3DStamped::X] = state2.vel_linear.x();
+  velocity_linear2->data()[fuse_variables::VelocityLinear3DStamped::Y] = state2.vel_linear.y();
+  velocity_linear2->data()[fuse_variables::VelocityLinear3DStamped::Z] = state2.vel_linear.z();
+
+  velocity_angular2->data()[fuse_variables::VelocityAngular3DStamped::ROLL] =
+    state2.vel_angular.x();
+  velocity_angular2->data()[fuse_variables::VelocityAngular3DStamped::PITCH] =
+    state2.vel_angular.y();
+  velocity_angular2->data()[fuse_variables::VelocityAngular3DStamped::YAW] = state2.vel_angular.z();
+
+  acceleration_linear2->data()[fuse_variables::AccelerationLinear3DStamped::X] =
+    state2.acc_linear.x();
+  acceleration_linear2->data()[fuse_variables::AccelerationLinear3DStamped::Y] =
+    state2.acc_linear.y();
+  acceleration_linear2->data()[fuse_variables::AccelerationLinear3DStamped::Z] =
+    state2.acc_linear.z();
+
+  state1.position_uuid = position1->uuid();
+  state1.orientation_uuid = orientation1->uuid();
+  state1.vel_linear_uuid = velocity_linear1->uuid();
+  state1.vel_angular_uuid = velocity_angular1->uuid();
+  state1.acc_linear_uuid = acceleration_linear1->uuid();
+
+  state2.position_uuid = position2->uuid();
+  state2.orientation_uuid = orientation2->uuid();
+  state2.vel_linear_uuid = velocity_linear2->uuid();
+  state2.vel_angular_uuid = velocity_angular2->uuid();
+  state2.acc_linear_uuid = acceleration_linear2->uuid();
+
+  state_history_.emplace(beginning_stamp, std::move(state1));
+  state_history_.emplace(ending_stamp, std::move(state2));
+
+  // Scale process noise covariance pose by the norm of the current state twist
+  auto process_noise_covariance = process_noise_covariance_;
+  if (scale_process_noise_) {
+    common::scaleProcessNoiseCovariance(
+      process_noise_covariance, state1.vel_linear, state1.vel_angular,
+      velocity_linear_norm_min_, velocity_angular_norm_min_);
+  }
+
+  // Validate
+  process_noise_covariance *= dt;
+
+  if (!disable_checks_) {
+    try {
+      validateMotionModel(state1, state2, process_noise_covariance);
+    } catch (const std::runtime_error & ex) {
+      RCLCPP_ERROR_STREAM_THROTTLE(
+        logger_, *clock_, 10.0 * 1000,
+        "Invalid '" << name_ << "' motion model: " << ex.what());
+      return;
+    }
+  }
```

Comment:
```suggestion
  // Scale process noise covariance pose by the norm of the current state twist
  auto process_noise_covariance = process_noise_covariance_;
  if (scale_process_noise_) {
    common::scaleProcessNoiseCovariance(
      process_noise_covariance, state1.vel_linear, state1.vel_angular,
      velocity_linear_norm_min_, velocity_angular_norm_min_);
  }

  // Validate
  process_noise_covariance *= dt;

  if (!disable_checks_) {
    try {
      validateMotionModel(state1, state2, process_noise_covariance);
    } catch (const std::runtime_error & ex) {
      RCLCPP_ERROR_STREAM_THROTTLE(
        logger_, *clock_, 10.0 * 1000,
        "Invalid '" << name_ << "' motion model: " << ex.what());
      return;
    }
  }
  
  state_history_.emplace(beginning_stamp, std::move(state1));
  state_history_.emplace(ending_stamp, std::move(state2));
```
I believe `state1` and `state2` are being used after move here. Does switching the order still work? Seems like it should since none of the stuff after relies on `state_history_` as far as I can tell.

Reply:
Yes, it's true that in this case `state1` and `state2` are being used after move. I basically copied the implementation of `generateMotionModel` of the `Unicycle2D` class, where it's done in the same way.
Anyway, after some testing I found that the two object member values are not modified after a call to `std::move()`, I believe that this is since `state1` and `state2` are of `StateHistoryElement` struct type that:
- does not implement a custom move constructor
- has member types that recall to primitive types such as float, double (I'm not entirely sure about UUIDs which are `boost::uuids::uuid` but they behave in the same way), which remain unchanged after move

For example, I tried adding another `std::string` member to `StateHistoryElement` struct, assign a value to it, and I checked that after move it is left in valid and uninitialized state (an empty string in this case).

About your suggested changes, the only thing that I am not sure of is that in this case, if `validateMotionModel` fails than the two states are not inserted in the state history, while this happens in the current implementation. I don't know the exact consequences of this.
Maybe @svwilliams can you helps us here? Thank you!